python/taichi/lang/kernel_arguments.py

```diff
@@ -37,9 +37,35 @@ def extract(self, x):
 
 
 class ArgAnyArray:
-    """Type annotation for arbitrary arrays, including external arrays and Taichi ndarrays."""
+    """Type annotation for arbitrary arrays, including external arrays and Taichi ndarrays.
+
+    For external arrays, we can treat it as a Taichi field with Vector or Matrix elements by specifying element shape and layout.
+
+    Args:
+        element_shape (Tuple[Int], optional): () if scalar elements (default), (n) if vector elements, and (n, m) if matrix elements.
+        is_soa (bool, optional): Whether to change the layout from AOS (default) to SOA.
+    """
+    def __init__(self, element_shape=(), is_soa=False):
+        assert len(
```

Comment:
Raise errors instead of using assert. The idea is that the former handles user errors, while the later catches Taichi implementation bugs

Reply:
done